posts/2023-06-01-Rust-1.70.0.md

```diff
@@ -0,0 +1,74 @@
+
+If you'd like to help us out by testing future releases, you might consider updating locally to use the beta channel (`rustup default beta`) or the nightly channel (`rustup default nightly`). Please [report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you might come across!
+
+## What's in 1.70.0 stable
+
+### Sparse by default for crates.io
+
+Cargo's "sparse" protocol is now enabled by default for reading the index from crates.io. This feature was previously stabilized with [Rust 1.68.0](https://blog.rust-lang.org/2023/03/09/Rust-1.68.0.html#cargos-sparse-protocol), but still required configuration to use that with crates.io. The announced plan was to make that the default in 1.70.0, and here it is! (TODO: maybe a statement about effects from user and infra sides?)
+
+### `OnceCell` and `OnceLock`
+
+Two new types have been stabilized for one-time initialization of shared data, `OnceCell` and its thread-safe counterpart `OnceLock`. These can be used anywhere that immediate construction is not wanted, and perhaps not even possible like non-`const` data in global variables.
+
+```rust
+use std::sync::OnceLock;
+
+fn main() {
+    let winner_lock = OnceLock::new();
+
+    let winner = std::thread::scope(|s| {
+        s.spawn(|| winner_lock.set("thread"));
+
+        std::thread::yield_now(); // give them a chance...
+
+        winner_lock.get_or_init(|| "main")
+    });
+
+    println!("{winner} wins!");
+}
+```
+
+Crates such as `lazy_static` and `once_cell` have filled this need in the past, but now these building blocks are part of the standard library, ported from `once_cell`'s `unsync` and `sync` modules. There are still more methods that may be stabilized in the future, as well as companion `LazyCell` and `LazyLock` types that store their initializing function, but this first step in stabilization should already cover many use cases.
+
+### Named levels of debug information
+
+The `-Cdebuginfo` compiler option has previously only supported numbers 0..=2 for increasing amounts of debugging information, where Cargo defaults to 2 in dev and test profiles and 0 in release and bench profiles. These debug levels can now be set by name: "none" (0), "limited" (1), and "full" (2), as well as two new levels, "line-directives-only" and "line-tables-only".
+
+The Cargo and rustc documentation both called level 1 "line tables only" before, but it was more than that with information about all functions, just not types and variables. That level is now called "limited", and the new "line-tables-only" level is further reduced to the minimum needed for backtraces with filenames and line numbers. This may eventually become the level used for `-Cdebuginfo=1`. The other `line-directives-only` level is intended for NVPTX profiling, and is otherwise not recommended.
+
+### Enforced stability in the `test` CLI
```

Comment:
This change is confirmed to have broken certain test-related functionality in editors/IDEs like IntelliJ and VS Code, and I think that it would be worth mentioning explicitly; many a user of said functionality is likely completely unaware that the editors were reliant upon the prior libtest behavior

Reply:
I think https://github.com/rust-lang/rust/pull/109044 has started collecting respective citations. I could take a pass at suggested text later today, or at least within the next couple days. However, it might make sense for someone to try to pull together something more holistic; I'm unfamiliar with any plans or recommendations the IntelliJ Rust folks may have, similar for cargo2junit, etc.

I know I'm probably just on my soap box at this point, but I feel like this is a significant change which has a larger blast radius than initially believed, and that it will negatively impact the developer experience for many. Perhaps it would even be better discussed in more detail in a blog post and then linked from the relnotes 🤷

Reply:
cc @nrc @vlad20012 in case there's anything you'd recommend mentioning in the relnotes on the libtest changes

cc @epage in case there's anything you'd recommend mentioning based on any plans stemming from recent discussions with libs

Reply:
For my part, its too early / nebulous for affecting the discussion of release notes.

Reply:
urgh, I completely dropped the ball here - I was going to try and put together a backout PR for this and got assent from the libs API team to have a deprecation period. Anyway, I didn't and it's too late now.

I think it is worth mentioning that at least cargo2junit and intellij will break, and to see those places for instructions on how to fix it

Reply:
I would be grateful if you would mention the fact that IntelliJ Rust uses `--format json` option in order to show a test tree in the test tool window, and this functionality got broken because of this change, and also if you would mention the tracking issue for `--format json`: https://github.com/rust-lang/rust/issues/49359.

Reply:
For reference, here's @vlad20012's change in IntelliJ Rust: https://github.com/intellij-rust/intellij-rust/pull/10511

I've added a brief note about these breakages and a link to that tracking issue.